File: app/src/main/java/com/cottondroid/todoachieved/task/model/TodoTaskDao.java
package com.cottondroid.todoachieved.task.model;

import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;
import java.util.concurrent.Callable;

import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Single;

@Dao
public abstract class TodoTaskDao {
    @Query("SELECT * FROM todotask ORDER BY createdDate")
    public abstract Flowable<List<TodoTask>> list();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract long insertSync(TodoTask task);

    public Single<Long> insertOrReplace(final TodoTask todoTask) {
        return Single.fromCallable(new Callable<Long>() {
            @Override
            public Long call() throws Exception {
                if (todoTask.getId() == null && todoTask.getServerId() != null) {
                    Long id = getIdByServerId(todoTask.getServerId());
                    todoTask.setId(id);
                }
                return insertSync(todoTask);
            }
        });
    }

    @Update
    public abstract int updateSync(TodoTask task);

    public Single<Integer> update(final TodoTask todoTask) {
        return Single.fromCallable(new Callable<Integer>() {
            @Override
            public Integer call() throws Exception {
                if (todoTask.getId() == null) {
                    Long id = getIdByServerId(todoTask.getServerId());
                    todoTask.setId(id);
                }
                return updateSync(todoTask);
            }
        });
    }

    @Delete
    public abstract int deleteSync(TodoTask task);

    public Single<Integer> delete(final TodoTask todoTask) {
        return Single.fromCallable(new Callable<Integer>() {
            @Override
            public Integer call() throws Exception {
                if (todoTask.getId() == null) {
                    Long id = getIdByServerId(todoTask.getServerId());
                    todoTask.setId(id);
                }
                return deleteSync(todoTask);
            }
        });
    }

    @Query("SELECT id FROM todotask WHERE serverId = :serverId")
    public abstract Long getIdByServerId(String serverId);

    @Query("SELECT * FROM todotask WHERE id = :taskId")
    public abstract Single<TodoTask> load(long taskId);

    @Query("SELECT serverId FROM todotask ORDER BY serverCreatedTimestamp desc LIMIT 1")
    public abstract Maybe<String> getLastServerId();

}

File: app/src/main/java/com/cottondroid/todoachieved/task/model/TodoTaskDao.kt
package com.cottondroid.todoachieved.task.model

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.Single

@Dao
abstract class TodoTaskDao {
    @Query("SELECT * FROM todotask ORDER BY createdDate")
    abstract fun list(): Flowable<List<TodoTask>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertSync(task: TodoTask): Long

    fun insertOrReplace(todoTask: TodoTask): Single<Long> {
        return Single.fromCallable {
            insertSync(
                    if (todoTask.id == null && todoTask.serverId != null) {
                        val id = getIdByServerId(todoTask.serverId)
                        todoTask.copy(id = id)
                    } else {
                        todoTask
                    }
            )
        }
    }

    @Update
    abstract fun updateSync(task: TodoTask): Int

    fun update(todoTask: TodoTask): Single<Int> {
        return Single.fromCallable {
            updateSync(
                    if (todoTask.id == null) {
                        val id = getIdByServerId(todoTask.serverId!!)
                        todoTask.copy(id = id)
                    } else {
                        todoTask
                    }
            )
        }
    }

    @Delete
    abstract fun deleteSync(task: TodoTask): Int

    fun delete(todoTask: TodoTask): Single<Int> {
        return Single.fromCallable {
            deleteSync(todoTask)
        }
    }

    @Query("SELECT id FROM todotask WHERE serverId = :serverId")
    abstract fun getIdByServerId(serverId: String): Long?

    @Query("SELECT * FROM todotask WHERE id = :taskId")
    abstract fun load(taskId: Long): Single<TodoTask>

    @get:Query("SELECT serverId FROM todotask ORDER BY serverCreatedTimestamp desc LIMIT 1")
    abstract val lastServerId: Maybe<String>
}